main/src/ca/uwaterloo/flix/api/lsp/provider/CompletionProvider.scala

```diff
@@ -0,0 +1,454 @@
+/*
+ * Copyright 2022 Paul Butcher
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package ca.uwaterloo.flix.api.lsp.provider
+
+import ca.uwaterloo.flix.api.lsp._
+import ca.uwaterloo.flix.language.ast.{Ast, Symbol, Type, TypeConstructor, TypedAst}
+import ca.uwaterloo.flix.language.fmt.{Audience, FormatScheme, FormatType}
+import ca.uwaterloo.flix.language.phase.Parser.Letters
+import ca.uwaterloo.flix.language.phase.Resolver.DerivableSyms
+import ca.uwaterloo.flix.util.InternalCompilerException
+import org.json4s.JsonAST.JObject
+import org.json4s.JsonDSL._
+import org.parboiled2.CharPredicate
+
+/*
+ * CompletionProvider
+ *
+ * Takes a source file, along with the position of the cursor within that file, and returns a list of CompletionItems.
+ *
+ * https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_completion
+ *
+ * This list is not displayed to the user as-is, the client always both sorts and filters the list (or at least,
+ * VSCode does). Therefore we always need to provide both filterText (currently copied from label) and sortText.
+ *
+ * To ensure that completions are displayed "most useful" first, we preceed sortText with a number as follows:
+ *
+ * 1: High confidence completions (currently with and instance)
+ * 2: Def and Sig
+ * 5: Var
+ * 8: Snippets
+ * 9: Keywords
+ *
+ * Note that we use textEdit rather than insertText to avoid relying on VSCode's tokenisation, so we can ensure that
+ * we're consistent with Flix's parser.
+ */
+
+object CompletionProvider {
+  private implicit val audience: Audience = Audience.External
+
+  def autoComplete(uri: String, pos: Position, source: Option[String])(implicit index: Index, root: TypedAst.Root): JObject = {
+    val completions = source.flatMap(getContext(_, uri, pos)) match {
+      case None => Nil
+      case Some(context) => getCompletions()(context, index, root)
+    }
+
+    ("status" -> "success") ~ ("result" -> CompletionList(isIncomplete = true, completions).toJSON)
+  }
+
+  private def getCompletions()(implicit context: Context, index: Index, root: TypedAst.Root): List[CompletionItem] = {
```

Comment:
Add a comment that explains the order of completions is immaterial since we use the "priority trick" (or whatever we want to call it).

Reply:
Will do